corporate/lib/remote_billing_util.py

```diff
@@ -133,9 +134,8 @@ def get_remote_server_and_user_from_session(
     request: HttpRequest,
     server_uuid: str,
 ) -> Tuple[RemoteZulipServer, Optional[RemoteServerBillingUser]]:
-    identity_dict = cast(
-        Optional[LegacyServerIdentityDict],
-        get_identity_dict_from_session(request, realm_uuid=None, server_uuid=server_uuid),
+    identity_dict: Optional[LegacyServerIdentityDict] = get_identity_dict_from_session(
+        request, realm_uuid=None, server_uuid=server_uuid
```

Comment:
I don't understand why mypy keeps changing its mind on whether these need a `cast`, or `cast` is redundant and it wants me to simply declare the type like this. Ref https://github.com/zulip/zulip/pull/28100#discussion_r1421299464 in the previous PR

Mypy suggests using "tagged unions" to deal with `Union`s of `TypeDict`s (due to the inability to `assert isinstance`) - https://mypy.readthedocs.io/en/stable/literal_types.html#tagged-unions But adding an artificial `Literal` to those dicts also feels messy. I guess I should have used `dataclass` instead  to begin with perhaps

Reply:
I would expect the solution to look like a set of overloads that narrow the type for, say, `realm_uuid=None` meaning that you get back a `LegacyServerIdentityDict`, to be the right implementation for this.